modules/boltlib/lib/puppet/functions/run_task.rb

```diff
@@ -59,22 +52,26 @@ def run_task_instance(task, *targets)
       )
     end
 
+    use_args = task_args.nil? ? {} : task_args
+    task_signature.runnable_with?(use_args) do |mismatch|
+      raise Puppet::ParseError, mismatch
+    end || (raise Puppet::ParseError, 'Task parameters did not match')
+    task = task_signature.task
+
     # Ensure that that given targets are all Target instances
+    targets = [targets] unless targets.is_a?(Array)
     targets = targets.flatten.map { |t| t.is_a?(String) ? Puppet::Pops::Types::TypeFactory.target.create(t) : t }
     if targets.empty?
-      call_function('debug', "Simulating run of task #{task._pcore_type.name} - no targets given - no action taken")
-      Puppet::Pops::Types::ExecutionResult::EMPTY_RESULT
+      call_function('debug', "Simulating run of task #{task.name} - no targets given - no action taken")
+      Puppet::Pops::EMPTY_HASH
     else
-      # Awaits change in the executor, enabling it receive Target instances
+      # TODO: Awaits change in the executor, enabling it receive Target instances
       hosts = targets.map(&:host)
 
       # TODO: separate handling of default since it's platform specific
-      input_method = task._pcore_type['input_method'].value
+      input_method = task.input_method
 
-      arguments = task.task_args
-      Puppet::Pops::Types::ExecutionResult.from_bolt(
-        executor.run_task(executor.from_uris(hosts), task.executable_path, input_method, arguments)
-      )
+      executor.run_task(executor.from_uris(hosts), task.executable, input_method, use_args, &block)
```

Comment:
I might be wrong here but I was under the impression that we eventually wanted the `executor.run_task` to return an `ExecutionResult` so that it didn't need to get wrapped here. If so, that means that the handling of that `ExecutionResult` must be done above this function.

Regardless of that, I think it less ideal to have the function use different return types depending on how it is called. I would like this function to always return an `ExecutionResult`. It's in line with the intention to let a `run_plan` return an aggregated `ExecutionResult`. In order to support this, Bolt must be able to present an `ExecutionResult` to the user (same goes for `run_script`, `run_command`, and `file_upload`.

Reply:
While making the executor use the pcore types for its arguments & return value is the cleanest solution, it unfortunately makes us pay the price of loading puppet_pal even when executing simple scripts.
Are we OK with it, @adreyer?

Reply:
I don't think we can unless we can move ExecutionResult out of puppet or decrease the performance penalty of loading puppet. All of the run_* functions should return similar values and loading puppet for commands, scripts and files should be as fast as possible.

Reply:
@joshcooper , FYI and off-topic in this PR but here's one reason to rewrite Bolt in Go. There is a performance penalty caused by Puppet load times.

Reply:
@thallgren It seems premature to worry about puppet load times when you're making hundreds of network connections, performing SSH/WinRM+SSL authentication, and executing commands on remote systems. Do you know what percentage of time is taken for compilation and what the speedup would be?

Reply:
@joshcooper , I completely agree with that. But what about the case at hand? Where puppet is loaded just to run a task, a script, or a command?